File: tests/tests/ClickAndMoveTest/ClickAndMoveTest.cpp
#include "ClickAndMoveTest.h"
#include "../testResource.h"

enum
{
	kTagSprite = 1,
};

void ClickAndMoveTestScene::runThisTest()
{
    CCLayer* pLayer = new MainLayer();
    pLayer->autorelease();

    addChild(pLayer);
    CCDirector::sharedDirector()->replaceScene(this);
}

MainLayer::MainLayer()
{
	setIsTouchEnabled(true);
	
	CCSprite* sprite = CCSprite::spriteWithFile(s_pPathGrossini);
	
	CCLayer* layer = CCLayerColor::layerWithColor(ccc4f(255,255,0,255));
	addChild(layer, -1);
		
	addChild(sprite, 0, kTagSprite);
	sprite->setPosition( CCPointMake(20,150) );
	
	sprite->runAction( CCJumpTo::actionWithDuration(4, CCPointMake(300,48), 100, 4) );
	
	layer->runAction( CCRepeatForever::actionWithAction( 
														(CCActionInterval*)( CCSequence::actions(	
																			CCFadeIn::actionWithDuration(1),
																			CCFadeOut::actionWithDuration(1),
																			NULL) )
														) ); 
}

void MainLayer::ccTouchesEnded(CCSet *pTouches, CCEvent *pEvent)
{
    CCSetIterator it = pTouches->begin();
	CCTouch* touch = (CCTouch*)(*it);
	
	CCPoint location = touch->locationInView();
	CCPoint convertedLocation = CCDirector::sharedDirector()->convertToGL(location);

	CCNode* s = getChildByTag(kTagSprite);
	s->stopAllActions();
	s->runAction( CCMoveTo::actionWithDuration(1, CCPointMake(convertedLocation.x, convertedLocation.y) ) );
	float o = convertedLocation.x - s->getPosition().x;
	float a = convertedLocation.y - s->getPosition().y;
	float at = (float) CC_RADIANS_TO_DEGREES( atanf( o/a) );
	
	if( a < 0 ) 
	{
		if(  o < 0 )
			at = 180 + fabs(at);
		else
			at = 180 - fabs(at);	
	}
	
	s->runAction( CCRotateTo::actionWithDuration(1, at) );
}

File: samples/Lua/TestLua/Resources/luaScript/ClickAndMoveTest/ClickAndMoveTest.lua
local size = CCDirector:sharedDirector():getWinSize()
local layer = nil
local kTagSprite = 1

local function initWithLayer()
    local sprite = CCSprite:create(s_pPathGrossini)

    local bgLayer = CCLayerColor:create(ccc4(255,255,0,255))
    layer:addChild(bgLayer, -1)

    layer:addChild(sprite, 0, kTagSprite)
    sprite:setPosition(CCPointMake(20,150))

    sprite:runAction(CCJumpTo:create(4, CCPointMake(300,48), 100, 4))

    bgLayer:runAction(CCRepeatForever:create(CCSequence:createWithTwoActions(
                                                 CCFadeIn:create(1),
                                                 CCFadeOut:create(1))))

    local function onTouchEnded(x, y)
        local s = layer:getChildByTag(kTagSprite)
        s:stopAllActions()
        s:runAction(CCMoveTo:create(1, CCPointMake(x, y)))
        local posX, posY = s:getPosition()
        local o = x - posX
        local a = y - posY
        local at = math.atan(o / a) / math.pi * 180.0

        if a < 0 then
            if o < 0 then
                at = 180 + math.abs(at)
            else
                at = 180 - math.abs(at)
            end
        end
        s:runAction(CCRotateTo:create(1, at))
    end

    local function onTouch(eventType, x, y)
        if eventType == "began" then
            return true
        elseif eventType == "ended" then
            return onTouchEnded(x, y)
        end
    end

    layer:setTouchEnabled(true)
    layer:registerScriptTouchHandler(onTouch)

    return layer
end

--------------------------------
-- Click And Move Test
--------------------------------
function ClickAndMoveTest()
    cclog("ClickAndMoveTest")
    local scene = CCScene:create()
    layer = CCLayer:create()

    initWithLayer()

    scene:addChild(CreateBackMenuItem())
    scene:addChild(layer)

    return scene
end
